wire-ios-data-model/Source/Model/UserClient/UserClient.swift

```diff
@@ -597,21 +598,76 @@ public extension UserClient {
         }
     }
 
+    var proteusProvider: ProteusProviding? {
+        if let context = self.managedObjectContext?.zm_sync {
+            return ProteusProvider(context: context)
+        }
+        return nil
+    }
+
     /// Creates a session between the selfClient and the given userClient
     /// Returns false if the session could not be established
     /// Use this method only for the selfClient
-    func establishSessionWithClient(_ client: UserClient, usingPreKey preKey: String) -> Bool {
+    func establishSessionWithClient(_ client: UserClient,
+                                    usingPreKey preKey: String,
+                                    proteusProviding: ProteusProviding) -> Bool {
         guard isSelfClient(), let sessionIdentifier = client.sessionIdentifier else { return false }
 
+        return proteusProviding.perform { proteusService in
+            establishSession(through: proteusService,
+                             client: client,
+                             sessionId: sessionIdentifier,
+                             preKey: preKey
+            )
+        } withKeyStore: { keystore in
+            establishSession(through: keystore,
+                             client: client,
+                             sessionId: sessionIdentifier,
+                             preKey: preKey
+            )
+        }
+    }
+
+    func establishSessionWithClient(_ client: UserClient, usingPreKey preKey: String) -> Bool {
+        guard let proteusProvider = proteusProvider else {
+            return false
+        }
+        return establishSessionWithClient(client, usingPreKey: preKey, proteusProviding: proteusProvider)
+    }
+
+    private func establishSession(through proteusService: ProteusServiceInterface,
+                                  client: UserClient,
+                                  sessionId: EncryptionSessionIdentifier,
+                                  preKey: String
+    ) -> Bool {
+        do {
+            // TODO: check if we should delete session if it exists before creating new one
```

Comment:
From previous discussions we arrive at the conclusion that there is a valid case where you want to delete a session, such as resetting the session. It seems like this method serves two purposes depending on how it is called:

1. Establishing a new session for the first time
2. Resetting a session by deleting the old one and creating a new one, when called any time after the initial setup.

If this is true, then it would be better to split the responsibility over two separate methods. But as a first step we can simply preserve the existing logic, then refactor afterwards. What do you think?

Reply:
preserve and do it in a second time I guess. I don't know it would affect keystore for exmple

Reply:
> I don't know it would affect keystore for exmple

I'm not sure what you mean. It would affect the keystore? Or you're asking if it would affect the keystore?

Reply:
yeah that's my question, does this affect the keystore method? I guess no